pygeos/_geometry.pyx

```diff
@@ -27,6 +35,102 @@ from pygeos._pygeos_api cimport (
+    cdef GEOSCoordSequence *seq = NULL
+
+    # Cast input arrays and define memoryviews for later usage
+    coordinates = np.asarray(coordinates, dtype=np.float64)
+    if coordinates.ndim != 2:
+        raise TypeError("coordinates is not a two-dimensional array.")
+
+    indices = np.asarray(indices, dtype=np.intp)
+    if indices.ndim != 1:
+        raise TypeError("indices is not a one-dimensional array.")
+
+    if coordinates.shape[0] != indices.shape[0]:
+        raise ValueError("geometries and indices do not have equal size.")
+
+    cdef unsigned int dims = coordinates.shape[1]
+    if dims not in {2, 3}:
+        raise ValueError("coordinates should N by 2 or N by 3.")
+
+    if geometry_type not in {0, 1, 2}:
+        raise ValueError(f"Invalid geometry_type: {geometry_type}.")
+
+    if coordinates.shape[0] == 0:
+        # return immediately if there are no geometries to return
+        return np.empty(shape=(0, ), dtype=np.object_)
+
+    if np.any(indices[1:] < indices[:indices.shape[0] - 1]):
+        raise ValueError("The indices should be sorted.")  
+
+    cdef double[:, :] coord_view = coordinates
+    cdef np.intp_t[:] index_view = indices
+
+    # get the geometry count per collection
+    cdef unsigned int[:] coord_counts = np.bincount(indices).astype(np.uint32)
+
+    # The final target array
+    cdef Py_ssize_t n_geoms = coord_counts.shape[0]
+    result = np.empty(shape=(n_geoms, ), dtype=object)
+    cdef object[:] result_view = result
+
+    with get_geos_handle() as geos_handle:
+        for geom_idx in range(n_geoms):
+            geom_size = coord_counts[geom_idx]
+
+            # check if we need to close a linearring
+            if geometry_type == 2:
+                ring_closure = 0
+                for coord_idx in range(dims):
+                    if coord_view[idx, coord_idx] != coord_view[idx + geom_size - 1, coord_idx]:
+                        ring_closure = 1
+                        break
+
+            seq = GEOSCoordSeq_create_r(geos_handle, geom_size + ring_closure, dims)
+            for coord_idx in range(geom_size):
+                if GEOSCoordSeq_setX_r(geos_handle, seq, coord_idx, coord_view[idx, 0]) == 0:
+                    return
+                if GEOSCoordSeq_setY_r(geos_handle, seq, coord_idx, coord_view[idx, 1]) == 0:
+                    return
+                if dims == 3:
+                    if GEOSCoordSeq_setZ_r(geos_handle, seq, coord_idx, coord_view[idx, 2]) == 0:
+                        return
+                idx += 1
+
+            if geometry_type == 0:
+                geom = GEOSGeom_createPoint_r(geos_handle, seq)
+            elif geometry_type == 1:
+                geom = GEOSGeom_createLineString_r(geos_handle, seq)
+            elif geometry_type == 2:
+                if ring_closure == 1:
+                    if GEOSCoordSeq_setX_r(geos_handle, seq, geom_size, coord_view[idx - geom_size, 0]) == 0:
+                        return
```

Comment:
Does the coordinate sequence need to be cleaned up if there are errors on line 118, 120, or 123?  Once `GEOSGeom_createLinearRing_r` is called, it takes ownership and would get cleaned up properly.